xds/src/main/java/io/grpc/xds/ClientXdsClient.java

```diff
@@ -228,86 +238,259 @@ private static LdsUpdate processClientSideListener(Listener listener, boolean pa
       throw new ResourceInvalidException(
           "Could not parse HttpConnectionManager config from ApiListener", e);
     }
+    return LdsUpdate.forApiListener(parseHttpConnectionManager(
+        hcm, rdsResources, filterRegistry, parseHttpFilter, true /* isForClient */));
+  }
+
+  private LdsUpdate processServerSideListener(
+      Listener proto, Set<String> rdsResources, boolean parseHttpFilter)
+      throws ResourceInvalidException {
+    return LdsUpdate.forTcpListener(parseServerSideListener(
+        proto, rdsResources, tlsContextManager, filterRegistry, parseHttpFilter));
+  }
+
+  @VisibleForTesting
+  static EnvoyServerProtoData.Listener parseServerSideListener(
+      Listener proto, Set<String> rdsResources, TlsContextManager tlsContextManager,
+      FilterRegistry filterRegistry, boolean parseHttpFilter) throws ResourceInvalidException {
+    if (!proto.getTrafficDirection().equals(TrafficDirection.INBOUND)) {
+      throw new ResourceInvalidException(
+          "Listener " + proto.getName() + " with invalid traffic direction: "
+              + proto.getTrafficDirection());
+    }
+    if (!proto.getListenerFiltersList().isEmpty()) {
+      throw new ResourceInvalidException(
+          "Listener " + proto.getName() + " cannot have listener_filters");
+    }
+    if (proto.hasUseOriginalDst()) {
+      throw new ResourceInvalidException(
+          "Listener " + proto.getName() + " cannot have use_original_dst set to true");
+    }
+
+    String address = null;
+    if (proto.getAddress().hasSocketAddress()) {
```

Comment:
Should you generate `ResourceInvalidException` if this is not a `SocketAddress`? We obviously don't support Pipe and EnvoyInternalAddress so....

Reply:
That's not handled in the [original code](https://github.com/grpc/grpc-java/blob/814655cdde5797854289ce5c2ec3e7b80ce0cf44/xds/src/main/java/io/grpc/xds/EnvoyServerProtoData.java#L542-L556) though. I can add it if you want...

Reply:
The spec says
> the xDS-returned Listener must have an address that matches the listening address provided. The Listener's address would be a TCP SocketAddress with matching address and port_value. The XdsServer must be "not serving" if the address does not match.

Just from the perspective of responsibility isolation/modularization, the XdsClient shouldn't try to extract the address and port from the resource name that the LDS watcher is subscribing to. It should just return the Listener with name equal to what the watcher subscribed to. Then it's the server's responsibility to check the received Listener has address (with port) matching its listening address.

So the code here seems fine to me. It just parses the SocketAddress (including the port). If it is a Pipe or EnvoyInternalAddress, the address given to the server is `null` and the server should not use it.

Reply:
Sounds good....